sphinx/application.py

```diff
@@ -147,6 +147,7 @@ def __init__(self, srcdir: str | os.PathLike[str], confdir: str | os.PathLike[st
         self.verbosity = verbosity
         self.extensions: dict[str, Extension] = {}
         self.registry = SphinxComponentRegistry()
+        self.project_class = Project  # can be modified by extensions in setup
```

Comment:
Ughh, I feel this should be a method, like `set_project_class`, with some form of `override = False` argument, as we do for other methods.

otherwise we are on shaky ground, if we start endorsing extensions to be directly changing attributes  of `Sphinx` (which in general will completely break sphinx)
ideally all these attributes would already be readonly, but such is life

Reply:
I'd suggest adding a documentation entry telling that this feature is to be used at your own risk and that extensions should mention that they change the Project class in their documentation.

And yes, I'd agree that we could have a method instead with this `override` argument (instead of a property setter) and make the project class read-only but retrievable via a property.